docs/topics/equality.md

```diff
@@ -1,50 +1,132 @@
+operation and its negated counterpart `!=`.
 By convention, an expression like `a == b` is translated to:
 
 ```kotlin
 a?.equals(b) ?: (b === null)
 ```
 
-If `a` is not `null`, it calls the `equals(Any?)` function, otherwise (`a` is `null`) it checks that `b`
-is referentially equal to `null`.
+If `a` is not `null`, it calls the `equals(Any?)` function. Otherwise (`a` is `null`), it checks that `b`
+is referentially equal to `null`:
+
+```kotlin
+fun main() {
+    var a = "hello"
+    var b = "hello"
+    var c = null
+    var d = null
+    var e = d
+
+    println(a == b)
+    // true
+    println(a == c)
+    // false
+    println(c == e)
+    // true
+}
+```
+{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}
 
 Note that there's no point in optimizing your code when comparing to `null` explicitly:
 `a == null` will be automatically translated to `a === null`.
 
+In Kotlin, the `equals()` function is inherited by all classes from the `Any` class. By default, the `equals()` function 
+implements [referential equality](#referential-equality). However, classes in Kotlin can override the `equals()` 
+function to provide a custom equality logic and, in this way, implement structural equality.
+
+Value classes and data classes are two specific types in Kotlin that automatically override `equals()`. 
+Value classes override the `equals()` function to implement structural equality by default.
+
+Similarly, data classes override the `equals()` function to implement structural equality by default. However, if the
+`equals()` function is marked as `final` in the parent class, its behavior remains unchanged.
+
+Distinctly, non-data classes (those not declared with the `data` modifier) do not override the 
+`equals()` function by default. Instead, non-data classes implement referential equality behavior inherited from the `Any` class.
+To implement structural equality, non-data classes require a custom equality logic to override the `equals()` function.
+
 To provide a custom equals check implementation, override the
-[`equals(other: Any?): Boolean`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-any/equals.html) function.
-Functions with the same name and other signatures, like `equals(other: Foo)`, don't affect equality checks with
+[`equals(other: Any?): Boolean`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-any/equals.html) function:
+
+```kotlin
+class Point(val x: Int, val y: Int) {
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is Point) return false
+
+        // Compares properties for structural equality
+        return this.x == other.x && this.y == other.y
+    }
+}
+```
+> When overriding the equals() function, you should also override the [hashCode() function](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-any/hash-code.html#kotlin.Any$hashCode()) 
```

Comment:
```suggestion
> When overriding the equals() function, you should also override the [hashCode() function](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-any/hash-code.html) 
```
I believe the anchor is not needed - that page fits in one screen. Also it can become obsolete after migrating documentation to the new dokka.